### PR TITLE
Add buttons to manual rerun tasks, both boefjes or normalizers

### LINKED ISSUE #291: Option to restart a task on task page
_Please add one or more of the following labels to your issue:_
`frontend`

**Is your feature request related to a problem? Please describe.**
Sometimes you want a recent task to run again (the normalizer or the boefje (or bit)). Currently, you have to find the page for the specific object or navigate to the respective boefje (even if it is only the normalizer that needs re-run).

**Describe the solution you'd like**
A button next to every task on the task page (or a selector combined with a button) to restart that task (both for boefjes, normalizers and bits).

**Describe alternatives you've considered**
Navigating painfully through KAT.

### LINKED ISSUE #1159: Rerun a (failed) boefje from the task list(s)
**Is your feature request related to a problem? Please describe.**
On the task list a Rerun this boefje button would be very useful, especially when the boefje has failed, and you want to try again.
All info (input ooi + boefje) is available, thus it should be failry simple.

**Describe the solution you'd like**
A column with a button to restart a given boefje+ooi combination

### LINKED ISSUE #1160: Rerun a (failed) normalizer run.
Is your feature request related to a problem? Please describe.
On the task list a Rerun this normalizer button would be very useful, especially when the normalizer has failed, and you want to try again.
All info (input raw file + normalizer) is available, thus it should be fairly simple.

Describe the solution you'd like
A column with a button to restart a given normalizer+raw-file combination

### PR DESCRIPTION
**Pull request by TwistMeister**
### Changes
Add buttons to restart (failed) tasks of boefjes and normalizers.

### Issue link
Closes #291 
Closes #1159 
Closes #1160 

### Proof
note: failed tasks was due to bytes/rabbitmq issues
Screenshot 2023-07-05 at 10 58 42
Screenshot 2023-07-05 at 10 58 48

---

### Code Checklist
- [x] All the commits in this PR are properly PGP-signed and verified;
- [x] This PR only contains functionality relevant to the issue; tickets have been created for newly discovered issues.
- [ ] I have written unit tests for the changes or fixes I made.
- [ ] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

### Communication
- [ ] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [ ] I have made corresponding changes to the documentation, if necessary.

---
## Checklist for code reviewers:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---
## Checklist for QA:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

File: rocky/katalogus/views/mixins.py [hunk 1]
```diff
@@ -60,6 +60,22 @@ def is_required_field(self, field: str) -> bool:
         return self.plugin_schema and field in self.plugin_schema.get("required", [])
 
 
+class NormalizerMixin(OctopoesView):
+    """
+    When a user wants to scan one or multiple OOI's,
+    this mixin provides the methods to construct the boefjes for the OOI's and run them.
```
Comment (underdarknl) [inline rocky/katalogus/views/mixins.py:66]: The comment still talks about boefjes, not normalizers

Comment (TwistMeister) [reply rocky/katalogus/views/mixins.py]: @underdarknl Apparently I forgot to save the file after I changed this, so it didn't get committed and pushed... It's pushed now though.

**Review by Rieven — CHANGES_REQUESTED** [review]
Looks good! A few things to consider:
- Add a column name "Action" for the td's
- What if there are no tasks? 
Do not show action button

**Comment by TwistMeister** [comment]
@Rieven `What if there are no tasks?` - When there are not tasks the button isn't rendered because no tasks row is rendered in the table, at all:

image

**Comment by TwistMeister** [comment]
@underdarknl update the logic for it to work with normalizers as well. Please review again :)

**Comment by Darwinkel** [comment]
## Checklist for QA:
- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [x] I confirmed that there are no unintended functional regressions in this branch:
    - [x] I have managed to pass the onboarding flow
    - [x] Objects and Findings are created properly
    - [x] Tasks are created and completed properly
- [x] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.

### What works:
* Restarting boefjes (triggers new boefje run with new metadata)
* Restarting normalizers (metadata identical to the previous run, which I suppose makes sense given that the normalizer's input is identical)

### Bug or feature?:
* There currently is no rate limiting: spamming F5 fills up the queue and task list.

File: rocky/katalogus/views/mixins.py [hunk 2]
```diff
@@ -60,6 +60,22 @@ def is_required_field(self, field: str) -> bool:
         return self.plugin_schema and field in self.plugin_schema.get("required", [])
 
 
+class NormalizerMixin(OctopoesView):
```
Comment (ammar92) [inline rocky/katalogus/views/mixins.py:63]: Why does it extend `OctopoesView`? It doesn't seem to depend on it.

It's also unusual to have mixins depending on views, its usually the other way around.

**Review by ammar92 — COMMENTED** [review]
Looks good in general, but we really need an automated test for this

File: rocky/rocky/views/tasks.py [hunk 3]
```diff
@@ -73,6 +80,31 @@ def get_queryset(self):
...
+            p_item = task.p_item
+            p_item.id = new_id
+            p_item.data.id = new_id
+
+            client.push_task(f"{task.type}-{self.organization.code}", task.p_item)
+
+            success_message = (
+                "Your task is running successfully in the background. \n "
+                "Results will be added to the object list when they are in. "
+                "It may take some time, a refresh of the page may be needed to show the results."
```
Comment (dekkers) [inline rocky/rocky/views/tasks.py:104]: The new task is scheduled, but doesn't have to be running the background yet.

File: rocky/rocky/templates/tasks/boefjes.html [hunk 4]
```diff
@@ -28,6 +28,7 @@ <h1>{% translate "Boefjes" %}</h1>
                                     <th scope="col">{% translate "Status" %}</th>
                                     <th scope="col">{% translate "Created date" %}</th>
                                     <th scope="col">{% translate "Download task details" %}</th>
+                                    <th scope="col">{% translate "Restart" %}</th>
```
Comment (dekkers) [inline rocky/rocky/templates/tasks/boefjes.html:31]: Would `Rerun` or `Reschedule` be a better name than `Restart`?

**Review by dekkers — CHANGES_REQUESTED** [review]

File: rocky/katalogus/views/mixins.py (hunk 2 above)
Comment (TwistMeister) [reply rocky/katalogus/views/mixins.py:63]: I don't know why it's this way around. I copied this from `BoefjeMixin`, but it's indeed not needed for this `NormalizerMixin`

File: rocky/rocky/views/tasks.py [hunk 5]
```diff
@@ -73,6 +80,31 @@ def get_queryset(self):
             messages.add_message(self.request, messages.ERROR, error_message)
             return []
 
+    def post(self, request, *args, **kwargs):
+        if "action" in self.request.POST:
+            self.handle_page_action(request.POST["action"])
+        return self.get(request, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (TwistMeister) [reply rocky/rocky/views/tasks.py:86]: Thanks! This solved it :)

File: rocky/rocky/views/tasks.py (hunk 3 above)
Comment (TwistMeister) [reply rocky/rocky/views/tasks.py:115]: Fixed

**Review by Rieven — APPROVED** [review]

**Review by underdarknl — APPROVED** [review]

**Pull request merged**